File: src/classes/BitReader.js
class BitReader {
    constructor(data) {
        this.data = data;
        this.bitIndex = 0;
    }

    readBits(count) {
        let result = 0;
        for (let i = 0; i < count; i++) {
            const byteIndex = Math.floor(this.bitIndex / 8);
            const bitOffset = 7 - (this.bitIndex % 8);
            if (byteIndex >= this.data.length) {
                throw new Error("Not enough data to read");
            }
            const bit = (this.data[byteIndex] >> bitOffset) & 1;
            result = (result << 1) | bit;
            this.bitIndex += 1;
        }
        return result;
    }

    alignToByte() {
        const remainder = this.bitIndex % 8;
        if (remainder !== 0) {
            this.bitIndex += 8 - remainder;
        }
    }

    readString() {
        this.alignToByte();
        const length = this.readBits(16);
        const resultBytes = Buffer.alloc(length);
        for (let i = 0; i < length; i++) {
            resultBytes[i] = this.readBits(8);
        }
        try {
            return resultBytes.toString("utf-8");
        } catch (e) {
            return resultBytes.toString("latin1");
        }
    }

    readMethod4() {
        const n = this.readBits(4);
        const bitCount = (n + 1) << 1;
        return this.readBits(bitCount);
    }

    readMethod393() {
        return this.readBits(8);
    }

    readMethod6(bitCount) {
        return this.readBits(bitCount);
    }
}

export { BitReader };

File: src/classes/bit_reader.py
class BitReader:
    def __init__(self, data):
        self.data = data
        self.bit_index = 0
    
    def read_bits(self, count):
        result = 0
        for i in range(count):
            byte_index = self.bit_index // 8
            bit_offset = 7 - (self.bit_index % 8)
            
            if byte_index >= len(self.data):
                raise Exception("Not enough data to read")
                
            bit = (self.data[byte_index] >> bit_offset) & 1
            result = (result << 1) | bit
            self.bit_index += 1
            
        return result
    
    def align_to_byte(self):
        remainder = self.bit_index % 8
        if remainder != 0:
            self.bit_index += 8 - remainder
    
    def read_string(self):
        self.align_to_byte()
        length = self.read_bits(16)
        
        result_bytes = bytearray(length)
        for i in range(length):
            result_bytes[i] = self.read_bits(8)
            
        try:
            return result_bytes.decode('utf-8')
        except Exception:
            return result_bytes.decode('latin1')
    
    def read_method4(self):
        n = self.read_bits(4)
        bit_count = (n + 1) << 1
        return self.read_bits(bit_count)
    
    def read_method393(self):
        return self.read_bits(8)
    
    def read_method6(self, bit_count):
        return self.read_bits(bit_count)